File: mcp-session-intel.py
"""
Session Intelligence MCP Server for OpenCode
Provides session analytics and pattern detection.
"""
import sys
import os
import json
from datetime import datetime
from pathlib import Path

if sys.platform == "win32":
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

MAHI_ROOT = Path(__file__).parent
sys.path.insert(0, str(MAHI_ROOT))

from command_center import SessionIntelligence

si = SessionIntelligence()


def analyze_patterns() -> dict:
    return {"timestamp": datetime.now().isoformat(), "analysis": si.analyze_patterns()}


def get_recommendations() -> list:
    analysis = si.analyze_patterns()
    return [a["action"] for a in analysis.get("next_actions", [])]


def handle_request(request: dict) -> dict:
    method = request.get("method", "")
    params = request.get("params", {})
    req_id = request.get("id")

    if method == "initialize":
        return {
            "jsonrpc": "2.0", "id": req_id,
            "result": {
                "protocolVersion": "2024-11-05",
                "capabilities": {"tools": {"listChanged": False}},
                "serverInfo": {"name": "session-intelligence", "version": "2.0.0"},
            },
        }
    elif method == "notifications/initialized":
        return None
    elif method == "tools/list":
        return {
            "jsonrpc": "2.0", "id": req_id,
            "result": {"tools": [
                {"name": "session_analyze", "description": "Analyze session patterns", "inputSchema": {"type": "object", "properties": {}}},
                {"name": "session_recommendations", "description": "Get next action recommendations", "inputSchema": {"type": "object", "properties": {}}},
            ]},
        }
    elif method == "tools/call":
        tool_name = params.get("name", "")
        try:
            if tool_name == "session_analyze":
                result = analyze_patterns()
            elif tool_name == "session_recommendations":
                result = get_recommendations()
            else:
                return {"jsonrpc": "2.0", "id": req_id, "error": {"code": -32601, "message": f"Unknown tool: {tool_name}"}}
            return {"jsonrpc": "2.0", "id": req_id, "result": {"content": [{"type": "text", "text": json.dumps(result, indent=2, ensure_ascii=False, default=str)}]}}
        except Exception as e:
            return {"jsonrpc": "2.0", "id": req_id, "error": {"code": -32000, "message": str(e)}}
    return {"jsonrpc": "2.0", "id": req_id, "error": {"code": -32601, "message": f"Unknown method: {method}"}}


def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            request = json.loads(line)
            response = handle_request(request)
            if response is not None:
                print(json.dumps(response, ensure_ascii=False))
                sys.stdout.flush()
        except json.JSONDecodeError:
            pass


if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] != '{"jsonrpc"':
        cmd = sys.argv[1]
        if cmd == "analyze":
            print(json.dumps(analyze_patterns(), indent=2, ensure_ascii=False))
        elif cmd == "recommendations":
            for r in get_recommendations():
                print(f"  - {r}")
        else:
            print(f"Usage: python mcp-session-intel.py [analyze|recommendations]")
    else:
        main()

File: mcp-session-intel.js
#!/usr/bin/env node
/**
 * Session Intelligence MCP Server for OpenCode
 */
const fs = require("fs");
const path = require("path");

const STATE_FILE = path.join("C:", "Users", "Admin", "My Drive", "LifeWorkspace", ".session-state.json");
const SESSIONS_DIR = path.join("C:", "Users", "Admin", "My Drive", "LifeWorkspace", "15_Advanced_Tools", "sessions");

function loadState() {
  try { return JSON.parse(fs.readFileSync(STATE_FILE, "utf-8")); } catch { return {}; }
}

function loadSessions() {
  const indexFile = path.join(SESSIONS_DIR, "index.json");
  try { return JSON.parse(fs.readFileSync(indexFile, "utf-8")).sessions || []; } catch { return []; }
}

function analyze() {
  const state = loadState();
  const sessions = loadSessions();
  const projects = {};
  const decisions = state.recent_decisions || [];

  for (const s of sessions) {
    const p = s.active_project || "unknown";
    projects[p] = (projects[p] || 0) + 1;
  }

  return {
    timestamp: new Date().toISOString(),
    session_state_loaded: Object.keys(state).length > 0,
    archived_sessions: sessions.length,
    active_project: state.active_project || "none",
    projects,
    recent_decisions_count: decisions.length,
    pending_tasks: (state.pending_tasks || []).length,
  };
}

function recommendations() {
  const state = loadState();
  const recs = [];
  const tasks = state.pending_tasks || [];
  if (tasks.length > 0) recs.push(`You have ${tasks.length} pending tasks`);
  if (state.active_project) recs.push(`Active project: ${state.active_project}`);
  if (!state.last_session) recs.push("No recent session found");
  if (tasks.length === 0 && !state.active_project) recs.push("All clear — consider starting a new task");
  return recs;
}

const TOOLS = [
  { name: "session_analyze", description: "Analyze session patterns and detect trends", inputSchema: { type: "object", properties: {} } },
  { name: "session_recommendations", description: "Get recommendations based on session history", inputSchema: { type: "object", properties: {} } },
  { name: "session_state", description: "Get current session state", inputSchema: { type: "object", properties: {} } },
];

function handle(req) {
  const { method, params, id } = req;
  if (method === "initialize") return { jsonrpc: "2.0", id, result: { protocolVersion: "2024-11-05", capabilities: { tools: {} }, serverInfo: { name: "session-intel", version: "1.0.0" } } };
  if (method === "notifications/initialized") return null;
  if (method === "tools/list") return { jsonrpc: "2.0", id, result: { tools: TOOLS } };
  if (method === "tools/call") {
    const { name } = params;
    try {
      let result;
      if (name === "session_analyze") result = analyze();
      else if (name === "session_recommendations") result = recommendations();
      else if (name === "session_state") result = loadState();
      else return { jsonrpc: "2.0", id, error: { code: -32601, message: `Unknown tool: ${name}` } };
      return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: JSON.stringify(result, null, 2) }] } };
    } catch (e) { return { jsonrpc: "2.0", id, error: { code: -32000, message: e.message } }; }
  }
  return { jsonrpc: "2.0", id, error: { code: -32601, message: `Unknown method: ${method}` } };
}

process.stdin.setEncoding("utf-8");
let buf = "";
process.stdin.on("data", (chunk) => {
  buf += chunk;
  const lines = buf.split("\n");
  buf = lines.pop();
  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    try { const req = JSON.parse(trimmed); const resp = handle(req); if (resp) process.stdout.write(JSON.stringify(resp) + "\n"); } catch {}
  }
});
